src/Typin/Typin/Experimental/Design.cs

```diff
@@ -0,0 +1,174 @@
+        /// in a list of other sibling commands
+        /// </summary>
+        public string Summary { get; init; }
+
+        /// <summary>
+        /// A description, possibly more extense than the <see cref="Summary"/> decribing what the command does
+        /// </summary>
+        public string Description { get; init; }
+
+        /// <summary>
+        /// List of argument descriptors specifying the arguments of current command
+        /// </summary>
+        public IReadOnlyCollection<CommandArgumentDescriptor> Arguments { get; init; }
+
+        /// <summary>
+        /// List of option descriptors specifying the options of current command
+        /// </summary>
+        public IReadOnlyCollection<CommandOptionDescriptor> Options { get; init; }
+
+        /// <summary>
+        /// List of examples to show to the user for current command
+        /// </summary>
+        public IReadOnlyCollection<CommandExampleDescriptor> Examples { get; init; }
+    }
+
+    public class CommandArgumentDescriptor
+    {
+        public string Name { get; init; }
+        public string Summary { get; init; }
+        public string Description { get; init; }
+        public int Position { get; init; }
+        public bool Multiple { get; init; }
+    }
+
+    public enum OptionKind
+    {
+        Flag,
+        Value,
+    }
+
+    public class CommandOptionDescriptor
+    {
+        public string Name { get; init; }
+        public IReadOnlyCollection<string> ShortAliases { get; init; }
+
+        /// <summary>
+        /// For flags, you can specify a negative name, like no-verbose, no-build, etc., so the user can
+        /// explicitly indicate that he wants to give a false value for the flag, disregarding it's default value
+        /// </summary>
+        public string NegativeName { get; init; }
+        public string Summary { get; init; }
+        public string Description { get; init; }
+        public bool Multiple { get; init; }
+        public OptionKind OptionKind { get; init; }
+        public string DefaultValue { get; init; }
+    }
+
+    public class CommandExampleDescriptor
+    {
+        public string Description { get; init; }
+    }
+
+    // Pure Base Runtime
+
+    public interface ICommandRunner
+    {
+        ValueTask ExecuteAsync(CommandRunnerContext context, CancellationToken cancellationToken);
+    }
+
+    public class CommandRunnerContext
+    {
+        public ICliContext CliContext { get; init; }
+        public CommandDescriptor CommandDescriptor { get; init; }
+        public string CommandLine { get; init; }
+        public IReadOnlyCollection<ParsedArgument> ParsedArguments { get; init; }
+        public IReadOnlyCollection<ParsedOption> ParsedOptions { get; init; }
+    }
+
+    public class ParsedArgument
+    {
+        public CommandArgumentDescriptor Descriptor { get; init; }
+        public int Position { get; init; }
+        public string Value { get; init; }
+    }
+
+    public class ParsedOption
```

Comment:
Sorry, I can hope for a better understanding after a couple of iterations. If you have any idea how to converge to less friction, lets talk about it.

Reply:
This is why I think that this changes are more aligned with a next version of the library, because I'm proposing a lot of changes. We can incorporate some of these to current 3.x version, and evaluate on that.

Reply:
You have amazingly thought-provoking and interesting ideas of new mechanisms. I'm thinking of writing a new concept/draft of dynamic commands myself to show my point of view, so that we can later combine what is best in our two visions. I think this will also help with disussion.

Reply:
That would be great! I prefer playing with the design until it clicks, and then start building.

Also I have shown a lot of mini features but lets focus on the feature at hand: Dynamic Commands